README.md

```diff
@@ -1,2 +1,195 @@
+
+#### Opaque tokens
+
+Token generators should know nothing about the request or context.
+
+#### Advanced Token Validation
+
+Tokens are layouted as `<key>.<signature>`. The following workflow requires an attacker to gain
+
+a. access to the database
+b. write permission in the persistent store,
+c. get hold of the token encryption secret.
+
+A database leak or (exclusively) the knowledge of the token encrpytion secret are not enough to maliciously obtain or create a valid token. Tokens and credentials can
+however still be stolen by man-in-the-middle attacks, by malicious or vulnerable clients and other attack vectors.
+
+**Issuance**
+
+1. The key is hashed using BCrypt (variable) and used as `<signature>`.
+2. The client is presented and entrusted with `<key>.<signature>` which can act for example as an access token or an authorize code.
+3. The signature is encrypted and stored in the database using AES (variable).
+
+**Validation**
+
+1. The client presents `<key>.<signature>`.
+2. It is validated if <key> matches <signature> using BCrypt (variable).
+3. The signature is encrypted using AES (variable).
+4. The encrypted signature is looked up in the database, failing validation if no such row exists.
+5. They key is considered valid and is now subject for other validations, like audience, redirect or state matching.
+
+A token generated by `generator.CryptoGenerator` looks like:
+
+```
+GUULhK6Od/7UAlnKvMau8APHSKXSRwm9aoOk56SHBns.JDJhJDEwJDdwVmpCQmJLYzM2VDg1VHJ5aEdVOE81NVdRSkt6bHBHR1QwOC9pbTNFWmpQRXliTWRPeDQy
+```
+
+#### Encrypt credentials at rest
+
+Credentials (token signatures, passwords and secrets) are always encrypted at rest.
+
+#### Implement peer reviewed IETF Standards
+
+Fosite implements [rfc6749](https://tools.ietf.org/html/rfc6749) and enforces countermeasures suggested in [rfc6819](https://tools.ietf.org/html/rfc6819).
+
+### Provide extensibility and interoperability
+
+... because OAuth2 is an extensible and flexible **framework**. Fosite let's you register new response types, new grant
+types and new response key value pares. This is useful, if you want to provide OpenID Connect on top of your
+OAuth2 stack. Or custom assertions, what ever you like and as long as it is secure. ;)
+
+## Usage
+
+This section is WIP and we welcome discussions via PRs or in the issues.
+
+### Store
+
+To use fosite, you need to implement `fosite.Storage`. Example implementations (e.g. postgres) of `fosite.Storage`
+will be added in the close future.
+
+### Authorize Endpoint
+
+```go
+package main
+
+import "github.com/ory-am/fosite"
+import "github.com/ory-am/fosite/session"
+import "github.com/ory-am/fosite/storage"
+
+// Let's assume that we're in a http handler
+func handleAuth(rw http.ResponseWriter, req *http.Request) {
+    oauth2 := fosite.NewDefaultOAuth2()
+
+    // ** This part of the API is not finalized yet **
+    // var store storage.Storage
+    // this should be something like: store := fosite.NewPostgreSQLStore()
```

Comment:
It may be useful to attach the storage to the oauth2 above ^
foside.NewDefaultOAuth2(store).
Then you can inject oauth2 into middleware context etc and it'll use a global storage.